### PR TITLE
feat(api): add application statistics

### LINKED ISSUE #2264: Add application statistics
## Description of the problem, including code/CLI snippet

The Gitlab provides some general statistics via `/application/statistics`.
See https://docs.gitlab.com/ee/api/statistics.html

## Expected Behavior

These data are available via `gitlab.statistics()`.

## Actual Behavior

Nothing.

## Specifications

  - python-gitlab version:
  - API version you are using (v3/v4):
  - Gitlab server version (or gitlab.com):

### PR DESCRIPTION
**Pull request by Shreya-7**
resolves #2264

### CODE REVIEW

**Comment by Shreya-7** [comment]
@nejch had a quick question. I've written a functional test for this functionality where I'm creating some entities (projects, groups, snippets, users, etc.) and then checking whether the statistics returned match or not (number of projects created = number from statistics).
However, this test keeps failing because there are a number of projects created in other functional tests that were not cleaned up, and hence those numbers are interfering with the test. Similarly for other entities.

I have two options here: either I go around modifying those tests to clean up any entities they create, or I do a count of created  (and undeleted) entities and just modify my test accordingly.
I would definitely prefer the first one, since it makes my test independent of others -- what do you think?

**Comment by lmilbaum** [comment]
> @nejch had a quick question. I've written a functional test for this functionality where I'm creating some entities (projects, groups, snippets, users, etc.) and then checking whether the statistics returned match or not (number of projects created = number from statistics). However, this test keeps failing because there are a number of projects created in other functional tests that were not cleaned up, and hence those numbers are interfering with the test. Similarly for other entities.
> 
> I have two options here: either I go around modifying those tests to clean up any entities they create, or I do a count of created (and undeleted) entities and just modify my test accordingly. I would definitely prefer the first one, since it makes my test independent of others -- what do you think?

Jumping in if I may. The fixture `reset_gitlab` can do the trick.

File: tests/unit/objects/test_statistics.py
Comment (lmilbaum) [inline tests/unit/objects/test_statistics.py:23]: Would it make sense to move the fixture code to conftest.py file?

File: tests/functional/api/test_statistics.py
Comment (lmilbaum) [inline tests/functional/api/test_statistics.py:6]: fixtures doesn't have to be imported. Just add the fixture to your function arguments and it will be executed.

File: tests/unit/objects/test_statistics.py
Comment (Shreya-7) [reply tests/unit/objects/test_statistics.py:23]: I'm not sure if any other unit tests will need this particular fixture, I'm following the same pattern as other unit tests

**Comment by Shreya-7** [comment]
> Jumping in if I may. The fixture `reset_gitlab` can do the trick.

Just what I was looking for, thanks!

File: tests/functional/api/test_statistics.py
Comment (Shreya-7) [reply tests/functional/api/test_statistics.py:5]: `reset_gitlab` isn't a fixture as of now - it's just a helper function as far as I can see. Not sure if we'd want to convert it into one as a function call is made to it within a different fixture.

File: tests/unit/objects/test_statistics.py
Comment (nejch) [reply tests/unit/objects/test_statistics.py:23]: I also think in this case it makes sense to keep the fixtures here, close to the scope where they are used as we do with other endpoint-specific fixtures:

**Comment by nejch** [comment]
Sorry @Shreya-7 I was away on a long weekend and catching up a bit!

> I have two options here: either I go around modifying those tests to clean up any entities they create, or I do a count of created (and undeleted) entities and just modify my test accordingly.
I would definitely prefer the first one, since it makes my test independent of others -- what do you think?

I would actually keep it even simpler than that, and only check that the attributes in the response can be successfully cast to an int but not actually assert on their values, that way we know the server response makes sense.

GitLab does a lot of creation and cleanup tasks asynchronously and I would not rely on absolute values here. We've actually tried to move away from that in some other tests - see https://github.com/python-gitlab/python-gitlab/pull/2118 for context.

That said, I see you've done a lot of work on getting `reset_gitlab` to behave. So we can keep that as it likely helps the testing in any case (in a separate commit maybe). WDYT?

File: tests/functional/conftest.py
Comment (nejch) [inline tests/functional/conftest.py:102]: This might be simpler, check if it has a parent:
```suggestion
        if group.parent_id:
```

**Comment by Shreya-7** [comment]
>I would actually keep it even simpler than that, and only check that the attributes in the response can be successfully cast to an int but not actually assert on their values, that way we know the server response makes sense.

This makes sense, let me try that!

> That said, I see you've done a lot of work on getting `reset_gitlab` to behave. So we can keep that as it likely helps the testing in any case (in a separate commit maybe). WDYT?

So should I open a new PR for this fix? I'm fine with it going in with this one, since my test does use this function

File: tests/functional/api/test_statistics.py
Comment (nejch) [reply tests/functional/api/test_statistics.py:5]: Agreed that this is a helper now, we could potentially move it to `tests.functional.helpers` later to avoid this confusion. Not needed for this PR though

File: tests/functional/api/test_statistics.py
Comment (nejch) [inline tests/functional/api/test_statistics.py:29]: Hehe just noticed the `hopefullythisdoesnthavetobeunique` ;) For what it's worth, with the current approach this entire block is not actually necessary. IMO you can remove it if you like, as GitLab should return these attributes even for a brand new instance without resources.

File: tests/functional/api/test_statistics.py
Comment (nejch) [inline tests/functional/api/test_statistics.py:35]: To make flake8 happy:

```suggestion
    assert statistics.snippets.isdigit()
    assert statistics.users.isdigit()
    assert statistics.groups.isdigit()
    assert statistics.projects.isdigit()
```

**Review by nejch — CHANGES_REQUESTED** [review]
@Shreya-7 up to you if you want this as a separate PR, I'm fine with it in here too! Just a few more nits from me here.

File: tests/functional/conftest.py
Comment (Shreya-7) [reply tests/functional/conftest.py]: yep, it's the problem of whoever wants to use this function next! After all the changes, just realised I spent more time on a function I ended up not using in my tests than on the actual feature lol :')

Comment (Shreya-7) [reply tests/functional/conftest.py:105]: actually, I had to revert this change as it started to inexplicably fail CLI functional tests.

Comment (nejch) [reply tests/functional/conftest.py:105]: Hehe @Shreya-7 thanks for your patience and sorry for the rollercoaster ride here, still good to see this can affect other tests in the future (whichever way it's done with subgroups) and we'll keep this in mind. Could you maybe just squash your PR to a single commit here for brevity in the release notes? I can do it too if you've had enough :sweat_smile:

Comment (Shreya-7) [reply tests/functional/conftest.py:105]: no worries! Do we want the commit message to include the reset_gitlab fix as well or just the feature?

Comment (nejch) [reply tests/functional/conftest.py]: There was a bit more work as there were some merge commits in between so I just force-pushed a squashed commit t o your branch @Shreya-7 ;) I'll just wait for CI to run now and otherwise LGTM!

**Review by nejch — APPROVED** [review]
Thanks for all the work here @Shreya-7! Hopefully next time it's less of a bumpy ride ;)

File: tests/functional/conftest.py
Comment (Shreya-7) [reply tests/functional/conftest.py]: Thanks so much!!!

**Pull request merged**
